### ISSUE TITLE
Moving Piece bug

### ISSUE THREAD

**Issue by derimmer** [vasl-developers/vasl#1736]
We are having an issue with some changes we introduced in ASLPieceMover to support being able to zoom boards independently of counters (or at least giving the appearance of doing so). 

As a result of these changes, in certain circumstances (as best as I can determine, this means adding a counter with an image larger than a hex), all subsequent counter movement is offset by a certain amount resulting in counters being in the wrong location on the board. 

This happens in VASL6.6.8 and VASSAL3.7.10+.

In the testing I have done, I have found two problems:

The first is that we are not getting any listeners working in ASLMap or ASLPieceMover that relate to piece movement. When I run the code in my dev environment and set breakpoints for mouse actions such as dropping, they only every occur in Map or PieceMover and never in ASLMap or ASLPieceMover. I have tried different forms of overriding methods and even replacing PieceMoveer wholus bolus with ASLPieceMover and it still end up invoking PieceMover and Map code. 

 The second is directly related to the problem we are having (I think!). 

PieceMover.AbstractDragHandler contains a makeDragImageCursorCommon() method which, amongst other things, sets the values of AbstractDragHandler.currentPieceOffsetX and AbstractDragHandler.currentPieceOffsetY. This method is not called when moving normal-sized methods around a map. When adding a counter with a larger image than normal (HexGrid, FFE), this method gets called and the values set. Subsequently when normal counters are moved, this method is not reached and the currentPieceOffsetX and currentPieceOffsetY value are not being reset. These values are used in PieceMover.AbstractDragHandler.drop() to position the moved Piece on the map and are causing the apparent "jump" when a dragged counter is dropped.

Next step is to set up a dev environment for 667 and determine what behaviour we had then at Drop time. Something has changed, I don't know what. 

@uckelman without asking you to debug this problem, would you have any thoughts about how to ensure that ASLMap and ASLPieceMover listeners are given priority over Map and PieceMover listeners?

**Comment by uckelman** [comment]
> The first is that we are not getting any listeners working in ASLMap or ASLPieceMover that relate to piece movement. When I run the code in my dev environment and set breakpoints for mouse actions such as dropping, they only every occur in Map or PieceMover and never in ASLMap or ASLPieceMover. 

I'd like to see a stack trace showing this.

For the other problem, it's hard to advise without knowing which commit is the breaking commit. The first thing I'd do is bisect to find it.

**Comment by derimmer** [comment]
Thanks! 

This is what I get from IntelliJ as a stack trace. Is this what you meant?

I put a breakpoint in PieceMover.AbstractDragHandler.drop() and this is what displays when the code execution hits the breakpoint after I drop a moved counter.

drop:2039, PieceMover$AbstractDragHandler (VASSAL.build.module.map)
drop:455, DropTarget (java.awt.dnd)
processDropMessage:548, SunDropTargetContextPeer (sun.awt.dnd)
dispatchDropEvent:864, SunDropTargetContextPeer$EventDispatcher (sun.awt.dnd)
dispatchEvent:788, SunDropTargetContextPeer$EventDispatcher (sun.awt.dnd)
dispatch:48, SunDropTargetEvent (sun.awt.dnd)
dispatchEventImpl:4860, Component (java.awt)
dispatchEventImpl:2324, Container (java.awt)
dispatchEvent:4827, Component (java.awt)
retargetMouseEvent:4948, LightweightDispatcher (java.awt)
processDropTargetEvent:4649, LightweightDispatcher (java.awt)
dispatchEvent:4511, LightweightDispatcher (java.awt)
dispatchEventImpl:2310, Container (java.awt)
dispatchEventImpl:2780, Window (java.awt)
dispatchEvent:4827, Component (java.awt)
dispatchEventImpl:775, EventQueue (java.awt)
run:720, EventQueue$4 (java.awt)
run:714, EventQueue$4 (java.awt)
executePrivileged:778, AccessController (java.security)
doPrivileged:400, AccessController (java.security)
doIntersectionPrivilege:87, ProtectionDomain$JavaSecurityAccessImpl (java.security)
doIntersectionPrivilege:98, ProtectionDomain$JavaSecurityAccessImpl (java.security)
run:747, EventQueue$5 (java.awt)
run:745, EventQueue$5 (java.awt)
executePrivileged:778, AccessController (java.security)
doPrivileged:400, AccessController (java.security)
doIntersectionPrivilege:87, ProtectionDomain$JavaSecurityAccessImpl (java.security)
dispatchEvent:744, EventQueue (java.awt)
pumpOneEventForFilters:203, EventDispatchThread (java.awt)
pumpEventsForFilter:124, EventDispatchThread (java.awt)
pumpEventsForHierarchy:113, EventDispatchThread (java.awt)
pumpEvents:109, EventDispatchThread (java.awt)
pumpEvents:101, EventDispatchThread (java.awt)
run:90, EventDispatchThread (java.awt)

**Comment by derimmer** [comment]
Here is the bisect I ran about two weeks ago. It pointed to the commit where we made changes to VASL code to support the BoardZoomer.vmdx extension.

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((vasl-6.6.8.vmod)|BISECTING)
$ git checkout v6.6.8-beta8
Previous HEAD position was 1887852a Merge pull request #1716 from vasl-developers/dev669
HEAD is now at d453f37d Merge pull request #1672 from vasl-developers/dev668

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((v6.6.8-beta8)|BISECTING)
$ git bisect good
Bisecting: 27 revisions left to test after this (roughly 5 steps)
[a2ed273ae026cf2dd841fad279afc00bd348ff15] Merge pull request #1695 from vasl-developers/dev668

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((a2ed273a...)|BISECTING)
$ git checkout vasl-6.6.8.vmod
Previous HEAD position was a2ed273a Merge pull request #1695 from vasl-developers/dev668
HEAD is now at 1887852a Merge pull request #1716 from vasl-developers/dev669

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((vasl-6.6.8.vmod)|BISECTING)
$ git bisect bad
Bisecting: 27 revisions left to test after this (roughly 5 steps)
[a2ed273ae026cf2dd841fad279afc00bd348ff15] Merge pull request #1695 from vasl-developers/dev668

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((a2ed273a...)|BISECTING)
$ git bisect bad
Bisecting: 12 revisions left to test after this (roughly 4 steps)
[2a65ae91a7c1372a16d9e5a5ef79896332c0a64c] #1692 new PF counter

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((2a65ae91...)|BISECTING)
$ git bisect bad
Bisecting: 6 revisions left to test after this (roughly 3 steps)
[13bae25e385622628af5f3b6bb51f4d9fe8d7caf] #1558 Pine Woods transformation

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((13bae25e...)|BISECTING)
$ git commit good
error: pathspec 'good' did not match any file(s) known to git

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((13bae25e...)|BISECTING)
$ git bisect good
Bisecting: 3 revisions left to test after this (roughly 2 steps)
[e615814911436b4db1decc46aee6c65e8c43527f] build 668beta9 with 379

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((e6158149...)|BISECTING)
$ git bisect good
Bisecting: 1 revision left to test after this (roughly 1 step)
[e49c52ce3662e7868c390153e18d43424ae4d75d] #1589 add Boardzoomer code changes

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((e49c52ce...)|BISECTING)
$ git bisect bad
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[f48b1bab3a326333b479f65766883e1b3e85d2ff] #1472 bug in previous commit

DougR@Doug MINGW64 /c/users/dougr/ideaprojects/vaslbisect ((f48b1bab...)|BISECTING)
$ git bisect good
e49c52ce3662e7868c390153e18d43424ae4d75d is the first bad commit
commit e49c52ce3662e7868c390153e18d43424ae4d75d
Author: DougR <[dougrimmer@rogers.com](mailto:dougrimmer@rogers.com)>
Date: Thu Mar 7 21:38:36 2024 -0500

#1589 add Boardzoomer code changes

src/VASL/build/module/ASLMap.java | 133 +--
src/VASL/build/module/map/ASLPieceMover.java | 1145 +++++++++++++++++++++++-
src/VASL/build/module/map/ASLStackMetrics.java | 106 ++-
src/VASL/build/module/map/BoardDataReader.java | 38 +
src/VASL/build/module/map/BoardZoomer.java | 182 ++++
5 files changed, 1540 insertions(+), 64 deletions(-)
create mode 100644 src/VASL/build/module/map/BoardZoomer.java

**Comment by uckelman** [comment]
My guess is that you're not getting the drag handler you want because of this line in `VASSAL.build.module.Map.addTo()`:
```
    theMap.setDropTarget(PieceMover.DragHandler.makeDropTarget(
      theMap, DnDConstants.ACTION_MOVE, this));
```
`ASLMap` doesn't override `addTo()`.

**Comment by derimmer** [comment]
A pretty damn good guess as usual!

I added addTo to ASLMap as follows:

> @Override
    public void addTo(Buildable b) {
        super.addTo(b);
        this.theMap.setDropTarget(ASLPieceMover.DragHandler.makeDropTarget(this.theMap, 2, this));
    }

When I ran VASL with this code (1) the ASLPieceMover.AbstractDragHandler.drop() method was called in lieu of the PieceMover one. Moreover, I have not been able to replicate the snap bug since then. I think we need to do more testing to be sure but it looks like we might have solved both problems.
